From 1f1a6824bc8991e2d0c42ed4ec6ebba30207a2a1 Mon Sep 17 00:00:00 2001
From: Brian Delwiche <delwiche@google.com>
Date: Tue, 26 May 2026 11:27:42 -0700
Subject: [PATCH 04/10] Fix heap buffer overflow in A2DP Opus decoder

The A2DP Opus decoder passes the buffer size in bytes to opus_decode()
where it expects samples-per-channel. This allows a crafted Opus
packet with a large frame duration to overflow the allocated buffer.

This change corrects the frame_size parameter passed to opus_decode()
by calculating the number of samples per channel that the buffer can
hold, assuming a fixed channel count of 2 as per the supported
configuration.

Additionally, this change:
- Removes a redundant loop that incorrectly decoded the same Opus
  packet multiple times when it contained multiple frames.
- Fixes incorrect logging of the number of frames and other packet
  metrics.

This patch was autogenerated by go/androidmendit.

Bug: 494914402
Bug: 503554165
#b/503554165
Test: atest net_test_stack_a2dp_codecs_native
Flag: EXEMPT BUGFIX

Change-Id: I3a1b150c0a761de54dedc5a27e0e9de96b19b16c
(cherry picked from commit e87caba4b8dfdb826c1d4b7cbb8a192d31ce7e85)
---
 system/stack/a2dp/a2dp_vendor_opus_decoder.cc | 65 ++++++++++---------
 system/stack/test/a2dp/a2dp_opus_unittest.cc  | 32 +++++++++
 2 files changed, 66 insertions(+), 31 deletions(-)

diff --git a/system/stack/a2dp/a2dp_vendor_opus_decoder.cc b/system/stack/a2dp/a2dp_vendor_opus_decoder.cc
index 4d2967c2439..527cd89814b 100644
--- a/system/stack/a2dp/a2dp_vendor_opus_decoder.cc
+++ b/system/stack/a2dp/a2dp_vendor_opus_decoder.cc
@@ -117,38 +117,41 @@ bool a2dp_vendor_opus_decoder_decode_packet(BT_HDR* p_buf) {
   numFrames = opus_packet_get_nb_frames(pBuffer, bufferSize);
   frameSize = opus_packet_get_samples_per_frame(pBuffer, A2DP_OPUS_CODEC_DEFAULT_SAMPLERATE);
   frameLen = opus_packet_get_nb_samples(pBuffer, bufferSize, A2DP_OPUS_CODEC_DEFAULT_SAMPLERATE);
-  uint32_t num_frames = pBuffer[0] & 0xf;
-
-  log::error("numframes {} framesize {} framelen {} bufferSize {}", num_frames, frameSize, frameLen,
-             bufferSize);
-  log::error("numChannels {} numFrames {} offset {}", numChannels, numFrames, p_buf->offset);
-
-  for (uint32_t frame = 0; frame < numFrames; ++frame) {
-    {
-      numChannels = opus_packet_get_nb_channels(pBuffer);
-
-      ret_val = opus_decode(a2dp_opus_decoder_cb.opus_handle,
-                            reinterpret_cast<unsigned char*>(pBuffer), bufferSize,
-                            a2dp_opus_decoder_cb.decode_buf, A2DP_OPUS_DECODE_BUFFER_LENGTH,
-                            0 /* flags */);
-
-      if (ret_val < OPUS_OK) {
-        log::error("Opus DecodeFrame failed {}, applying concealment", ret_val);
-        ret_val = opus_decode(a2dp_opus_decoder_cb.opus_handle, NULL, 0,
-                              a2dp_opus_decoder_cb.decode_buf, A2DP_OPUS_DECODE_BUFFER_LENGTH,
-                              0 /* flags */);
-      }
-
-      if (ret_val < OPUS_OK) {
-        log::error("Opus DecodeFrame retry failed with {}, dropping packet", ret_val);
-        return false;
-      }
-
-      size_t frame_len = ret_val * numChannels * sizeof(a2dp_opus_decoder_cb.decode_buf[0]);
-      a2dp_opus_decoder_cb.decode_callback(
-              reinterpret_cast<uint8_t*>(a2dp_opus_decoder_cb.decode_buf), frame_len);
-    }
+
+  if (numChannels != 1 && numChannels != 2) {
+    log::error("Invalid numChannels {} in Opus packet", numChannels);
+    return false;
+  }
+
+  log::info("numChannels={} numFrames={} frameSize={} frameLen={} bufferSize={} offset={}",
+               numChannels, numFrames, frameSize, frameLen, bufferSize, p_buf->offset);
+
+  ret_val = opus_decode(
+          a2dp_opus_decoder_cb.opus_handle, reinterpret_cast<unsigned char*>(pBuffer), bufferSize,
+          a2dp_opus_decoder_cb.decode_buf,
+          A2DP_OPUS_DECODE_BUFFER_LENGTH / (A2DP_OPUS_CODEC_OUTPUT_CHS * sizeof(int16_t)),
+          0 /* flags */);
+
+  if (ret_val < OPUS_OK) {
+    log::error("Opus DecodeFrame failed {}, applying concealment", ret_val);
+    ret_val = opus_decode(
+            a2dp_opus_decoder_cb.opus_handle, NULL, 0, a2dp_opus_decoder_cb.decode_buf,
+            A2DP_OPUS_DECODE_BUFFER_LENGTH / (A2DP_OPUS_CODEC_OUTPUT_CHS * sizeof(int16_t)),
+            0 /* flags */);
   }
+
+  if (ret_val < OPUS_OK) {
+    log::error("Opus DecodeFrame retry failed with {}, dropping packet", ret_val);
+    return false;
+  }
+
+  // The output size is determined by the codec channel selection, not the number of
+  // channels in the frame. Opus is always configured as stereo.
+  size_t frame_len =
+          ret_val * A2DP_OPUS_CODEC_OUTPUT_CHS * sizeof(a2dp_opus_decoder_cb.decode_buf[0]);
+  a2dp_opus_decoder_cb.decode_callback(reinterpret_cast<uint8_t*>(a2dp_opus_decoder_cb.decode_buf),
+                                       frame_len);
+
   return true;
 }
 
diff --git a/system/stack/test/a2dp/a2dp_opus_unittest.cc b/system/stack/test/a2dp/a2dp_opus_unittest.cc
index eb517fd9a84..c8b3f78df66 100644
--- a/system/stack/test/a2dp/a2dp_opus_unittest.cc
+++ b/system/stack/test/a2dp/a2dp_opus_unittest.cc
@@ -15,6 +15,8 @@
  */
 
 #include <gtest/gtest.h>
+#include <stdio.h>
+#include <opus.h>
 
 #include <chrono>
 #include <cstdint>
@@ -221,5 +223,35 @@ TEST_F(A2dpOpusTest, decoded_data_cb_invoked) {
   osi_free(packet);
 }
 
+TEST_F(A2dpOpusTest, decode_packet_60ms_overflow_repro) {
+  promise = {};
+  auto data_cb = +[](uint8_t* /*p_buf*/, uint32_t /*len*/) {};
+  InitializeDecoder(data_cb);
+
+  // Create a real 60ms Opus packet
+  int err;
+  OpusEncoder* enc = opus_encoder_create(48000, 2, OPUS_APPLICATION_VOIP, &err);
+  ASSERT_NE(enc, nullptr);
+  opus_encoder_ctl(enc, OPUS_SET_BITRATE(12000));
+
+  std::vector<int16_t> pcm_in(2880 * 2, 0);
+  std::vector<uint8_t> opus_pkt(4096);
+  int pkt_len = opus_encode(enc, pcm_in.data(), 2880, opus_pkt.data(), opus_pkt.size());
+  ASSERT_GT(pkt_len, 0);
+  opus_encoder_destroy(enc);
+
+  // A2DP Opus packet has 1-byte header
+  std::vector<uint8_t> data(pkt_len + 1);
+  data[0] = 1;  // 1 frame
+  std::copy(opus_pkt.begin(), opus_pkt.begin() + pkt_len, data.begin() + 1);
+
+  BT_HDR* packet = AllocateL2capPacket(data);
+  packet->offset = 0;
+
+  // This should trigger the overflow.
+  decoder_iface_->decode_packet(packet);
+  osi_free(packet);
+}
+
 }  // namespace testing
 }  // namespace bluetooth
-- 
2.55.0

